test/Base_Tests/src/Semantic/Private_Spec.enso

```diff
@@ -0,0 +1,76 @@
+
+from Standard.Base import all
+import Standard.Base.Errors.Common.Private_Access
+from Standard.Test import all
+from enso_dev.Helpers import Priv_Type
+
+polyglot java import org.enso.base_test_helpers.CallbackHelper
+
+
+add_specs spec_builder =
+    spec_builder.group "Private constructors" group_builder->
+        group_builder.specify "cannot directly call private constructor" <|
+            expect_access_error <| Priv_Type.Cons 42
+
+        group_builder.specify "can get reference to private constructor, but cannot call it" <|
+            cons_fn data = Priv_Type.Cons data
```

Comment:
It is not good that one can get reference to `private` constructor. One of the motivations to make something private is to refactor it, but when one changes name of such `private` constructor, one would break code that was previously OK obtaining the reference.

`private` constructors shall not be visible at all.

Reply:
I agree with your point. But I am afraid that at this point, access to a private constructor, or to any method, cannot be checked at compile time. This is because the current IR metadata do not contain sufficient information for that. For an example, see the following IR graph that uses `Priv_Type.Cons`:

![local Proj Main gv](https://github.com/enso-org/enso/assets/14013887/558d2e05-7d25-463c-bcca-41b4ba07b67b)

There is no `BindingsMap$Resolution` metadata for the `Cons` literal IR. It is difficult to add these metadata. There is an issue for that - https://github.com/enso-org/enso/issues/6729.

In conclusion, let's not add this functionality for now. I have, at least, added a comment explaining this in the test in 500a5e20db3822aa1ea3b7d19558f5fe2b33fd4d

Reply:
Can we at least get a runtime error when accessing `Cons`? E.g. prevent the constructor to be accessed at all?

Reply:
What do you mean by *accessing the constructor*? We get a runtime error as soon as the constructor is called. When you define a lambda function as `fn x = Priv_Type.Cons x`, the compiler cannot currently determine if a private constructor is called. At compilation time, we cannot ensure that a private constructor (or a method) is not referenced.

Reply:
I'd like to get an error when`fn x = [ Priv_Type.Cons, x]` is called. E.g. When other module tries to access `Priv_Type.Cons` without an attempt to call it.

Reply:
Unfortunatelly, this is not possible, as explained by my [other comment](https://github.com/enso-org/enso/pull/9692#issuecomment-2077585600). I have created a new issue to track this in https://github.com/enso-org/enso/issues/9794.